### PR TITLE
Add "Step" LR decay method backend by Pytorch

### PR DESCRIPTION
**Pull request by Jerry-Jzy**
## What is the purpose of this change:
Add "Step" LR decay method backend by Pytorch

### CODE REVIEW

File: deepxde/model.py [hunk 1]
```diff
@@ -39,6 +39,7 @@ def __init__(self, data, net):
 
         # Backend-dependent attributes
         self.opt = None
+        self.lr_scheduler = None
```
Comment (lululxvi) [inline deepxde/model.py:42]: Move this line after line 50
```python
elif backend_name == "pytorch":
    self.lr_scheduler = None
```

File: deepxde/optimizers/pytorch/optimizers.py [hunk 2]
```diff
@@ -31,11 +33,33 @@ def get(params, optimizer, learning_rate=None, decay=None):
     if learning_rate is None:
         raise ValueError("No learning rate for {}.".format(optimizer))
 
+    # Default optimizer is Adam
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:36]: No default optimizer.

File: deepxde/optimizers/pytorch/optimizers.py [hunk 3]
```diff
@@ -31,11 +33,33 @@ def get(params, optimizer, learning_rate=None, decay=None):
...
+    optim = torch.optim.Adam(params, lr=learning_rate)
+
+    if optimizer == "SGD":
+        optim = torch.optim.SGD(params, lr = learning_rate)
+
+    if optimizer == "RMSprop":
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:42]: `elif`

File: deepxde/optimizers/pytorch/optimizers.py [hunk 4]
```diff
@@ -31,11 +33,33 @@ def get(params, optimizer, learning_rate=None, decay=None):
...
+        optim = torch.optim.RMSprop(params,lr=learning_rate)
+
+    if optimizer not in ["SGD","RMSprop","adam"]:
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:45]: Why `adam` is not implemented?

File: deepxde/model.py (hunk 1 above)
Comment (Jerry-Jzy) [reply deepxde/model.py]: Fixed

File: deepxde/optimizers/pytorch/optimizers.py (hunk 2 above)
Comment (Jerry-Jzy) [reply deepxde/optimizers/pytorch/optimizers.py]: Fixed

File: deepxde/optimizers/pytorch/optimizers.py (hunk 3 above)
Comment (Jerry-Jzy) [reply deepxde/optimizers/pytorch/optimizers.py]: Fixed

File: deepxde/model.py [hunk 5]
```diff
@@ -49,6 +49,8 @@ def __init__(self, data, net):
             self.saver = None
         elif backend_name == "jax":
             self.opt_state = None
+        elif backend_name == "pytorch":
```
Comment (lululxvi) [inline deepxde/model.py:52]: swap the order of pytorch and jax

File: deepxde/optimizers/pytorch/optimizers.py [hunk 6]
```diff
@@ -31,11 +33,34 @@ def get(params, optimizer, learning_rate=None, decay=None):
...
         raise ValueError("No learning rate for {}.".format(optimizer))
 
+    if optimizer == "SGD":
+        optim = torch.optim.SGD(params, lr=learning_rate)
+    elif optimizer == "RMSprop":
+        optim = torch.optim.RMSprop(params, lr=learning_rate)
+    elif optimizer == "adam":
+        optim = torch.optim.Adam(params, lr=learning_rate)
+
+    if optimizer not in ["SGD", "RMSprop", "adam"]:
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:43]:
```
else:
    raise NotImplementedError(f"{optimizer} to be implemented for backend pytorch.")
```

File: deepxde/model.py (hunk 5 above)
Comment (Jerry-Jzy) [reply deepxde/model.py:50]: Fixed

File: deepxde/optimizers/pytorch/optimizers.py (hunk 6 above)
Comment (Jerry-Jzy) [reply deepxde/optimizers/pytorch/optimizers.py]: Fixed

File: deepxde/optimizers/pytorch/optimizers.py [hunk 7]
```diff
@@ -31,11 +33,33 @@ def get(params, optimizer, learning_rate=None, decay=None):
...
+        return torch.optim.lr_scheduler.StepLR(
+            optim, step_size=decay[1], gamma=decay[2]
         )
-    if optimizer == "adam":
-        return torch.optim.Adam(params, lr=learning_rate)
-    raise NotImplementedError(f"{optimizer} to be implemented for backend pytorch.")
+
+    # TODO: More learning rate scheduler
+    raise NotImplementedError(
+        f"{decay[0]} learning rate scheduler to be implemented for backend tensorflow."
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:64]: for backend tensorflow?

File: deepxde/optimizers/pytorch/optimizers.py [hunk 8]
```diff
@@ -31,11 +33,33 @@ def get(params, optimizer, learning_rate=None, decay=None):
...
+        lr_scheduler = None
+
+    return optim, lr_scheduler
+
+
+def _get_learningrate_scheduler(optim, decay):
+    if decay is None:
+        return None
+
+    if decay[0] == "Step":
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:57]: Step --> StepLR

File: deepxde/optimizers/pytorch/optimizers.py (hunk 7 above)
Comment (Jerry-Jzy) [reply deepxde/optimizers/pytorch/optimizers.py]: Fixed

File: deepxde/optimizers/pytorch/optimizers.py (hunk 8 above)
Comment (Jerry-Jzy) [reply deepxde/optimizers/pytorch/optimizers.py]: Fixed

File: deepxde/model.py [hunk 9]
```diff
@@ -47,9 +47,12 @@ def __init__(self, data, net):
         if backend_name == "tensorflow.compat.v1":
             self.sess = None
             self.saver = None
+        elif backend_name == "pytorch":
+            self.lr_scheduler = None
         elif backend_name == "jax":
             self.opt_state = None
 
+
```
Comment (lululxvi) [inline deepxde/model.py:55]: Delete this empty line.

File: deepxde/optimizers/pytorch/optimizers.py [hunk 10]
```diff
@@ -11,13 +11,15 @@ def is_external_optimizer(optimizer):
 
 def get(params, optimizer, learning_rate=None, decay=None):
     """Retrieves an Optimizer instance."""
+
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:14]: Delete this empty line.

File: deepxde/optimizers/pytorch/optimizers.py [hunk 11]
```diff
@@ -11,13 +11,15 @@ def is_external_optimizer(optimizer):
...
+    # Custom Optimizer
     if isinstance(optimizer, torch.optim.Optimizer):
-        return optimizer
+        optim = optimizer
 
     if optimizer in ["L-BFGS", "L-BFGS-B"]:
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:19]: elif

File: deepxde/optimizers/pytorch/optimizers.py [hunk 12]
```diff
@@ -31,11 +33,33 @@ def get(params, optimizer, learning_rate=None, decay=None):
     if learning_rate is None:
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:33]: else

File: deepxde/optimizers/pytorch/optimizers.py [hunk 13]
```diff
@@ -31,11 +33,33 @@ def get(params, optimizer, learning_rate=None, decay=None):
...
+    if optimizer == "SGD":
+        optim = torch.optim.SGD(params, lr=learning_rate)
+    elif optimizer == "RMSprop":
+        optim = torch.optim.RMSprop(params, lr=learning_rate)
+    elif optimizer == "adam":
+        optim = torch.optim.Adam(params, lr=learning_rate)
+    else:
+        raise NotImplementedError(f"{optimizer} to be implemented for backend pytorch.")
+
     if decay is not None:
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:45]: No need `if`, just use
```python
lr_scheduler = _get_learningrate_scheduler(optim, decay)
```

File: deepxde/model.py (hunk 9 above)
Comment (Jerry-Jzy) [reply deepxde/model.py]: Fixed

File: deepxde/optimizers/pytorch/optimizers.py (hunk 10 above)
Comment (Jerry-Jzy) [reply deepxde/optimizers/pytorch/optimizers.py]: Fixed

File: deepxde/optimizers/pytorch/optimizers.py (hunk 11 above)
Comment (Jerry-Jzy) [reply deepxde/optimizers/pytorch/optimizers.py]: Fixed

File: deepxde/optimizers/pytorch/optimizers.py (hunk 12 above)
Comment (Jerry-Jzy) [reply deepxde/optimizers/pytorch/optimizers.py:31]: Fixed

File: deepxde/optimizers/pytorch/optimizers.py (hunk 13 above)
Comment (Jerry-Jzy) [reply deepxde/optimizers/pytorch/optimizers.py]: Fixed

File: deepxde/optimizers/pytorch/optimizers.py [hunk 14]
```diff
@@ -30,12 +30,33 @@ def get(params, optimizer, learning_rate=None, decay=None):
 
     if learning_rate is None:
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:31]: Is this logic correct for L-BFGS with learning_rate as None?

Try an example with L-BFGS optimizer.

Comment (Jerry-Jzy) [reply deepxde/optimizers/pytorch/optimizers.py]: Fixed

File: deepxde/optimizers/pytorch/optimizers.py [hunk 15]
```diff
@@ -27,15 +27,35 @@ def get(params, optimizer, learning_rate=None, decay=None):
...
-        raise ValueError("No learning rate for {}.".format(optimizer))
-
-    if decay is not None:
-        # TODO: learning rate decay
-        raise NotImplementedError(
-            "learning rate decay to be implemented for backend pytorch."
+    else:
+        if learning_rate is None:
+            raise ValueError("No learning rate for {}.".format(optimizer))
+        elif optimizer == "SGD":
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:33]: "sgd"

File: deepxde/optimizers/pytorch/optimizers.py [hunk 16]
```diff
@@ -27,15 +27,35 @@ def get(params, optimizer, learning_rate=None, decay=None):
...
+                optim = torch.optim.SGD(params, lr=learning_rate)
+        elif optimizer == "RMSprop":
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:35]: rmsprop

File: deepxde/optimizers/pytorch/optimizers.py (hunk 15 above)
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:33]: if optimizer == "SGD":

Comment (Jerry-Jzy) [reply deepxde/optimizers/pytorch/optimizers.py]: Fixed

File: deepxde/optimizers/pytorch/optimizers.py (hunk 16 above)
Comment (Jerry-Jzy) [reply deepxde/optimizers/pytorch/optimizers.py]: Fixed

File: deepxde/optimizers/pytorch/optimizers.py (hunk 15 above)
Comment (Jerry-Jzy) [reply deepxde/optimizers/pytorch/optimizers.py]: Fixed

File: deepxde/optimizers/pytorch/optimizers.py [hunk 17]
```diff
@@ -27,15 +27,35 @@ def get(params, optimizer, learning_rate=None, decay=None):
...
-
-    if decay is not None:
-        # TODO: learning rate decay
-        raise NotImplementedError(
-            "learning rate decay to be implemented for backend pytorch."
+    else:
+        if learning_rate is None:
+            raise ValueError("No learning rate for {}.".format(optimizer))
+        if optimizer == "sgd":
+                optim = torch.optim.SGD(params, lr=learning_rate)
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:34]: indent is wrong

File: deepxde/optimizers/pytorch/optimizers.py [hunk 18]
```diff
@@ -27,15 +27,35 @@ def get(params, optimizer, learning_rate=None, decay=None):
...
+            optim = torch.optim.RMSprop(params, lr=learning_rate)
+        elif optimizer == "adam":
+            optim = torch.optim.Adam(params, lr=learning_rate)
+        else:
+            raise NotImplementedError(
+                f"{optimizer} to be implemented for backend pytorch."
+            )
+
+    lr_scheduler = _get_learningrate_scheduler(optim, decay)
+
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:45]: Delete this empty line.

File: deepxde/optimizers/pytorch/optimizers.py (hunk 17 above)
Comment (Jerry-Jzy) [reply deepxde/optimizers/pytorch/optimizers.py]: Fixed

File: deepxde/optimizers/pytorch/optimizers.py (hunk 18 above)
Comment (Jerry-Jzy) [reply deepxde/optimizers/pytorch/optimizers.py]: Fixed

**Pull request merged**